Scala 3 Regression: Unsupported source object for a case class that has a method with the same name as a field

@KacperFKorban I'm getting the following error in tapir after updating:

```
[error] -- Error: /Users/adamw/projects/tapir/docs/openapi-docs/src/test/scalajvm/sttp/tapir/docs/openapi/VerifyYamlTest.scala:847:15
[error] 847 |    docs.modify(_.paths.pathItems)
[error]     |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]     |Unsupported source object: must be a case class or sealed trait, but got: val <none>
```

Followed by a compiler crash:

```
[error] ## Exception when compiling 20 sources to /Users/adamw/projects/tapir/docs/openapi-docs/target/jvm-3/test-classes
[error] java.lang.AssertionError: assertion failed
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
[error] dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:2439)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3026)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3114)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:153)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedUnadapted(Inliner.scala:915)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3187)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3191)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3213)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3259)
[error] dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:1161)
[error] dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1165)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3058)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3115)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:153)
[error] dotty.tools.dotc.inlines.Inliner$InlineTyper.typedUnadapted(Inliner.scala:915)
(...)
```

The definition of `.pathItems` doesn't look extraordinary:

```
final case class Paths(
    pathItems: ListMap[String, PathItem] = ListMap.empty,
    extensions: ListMap[String, ExtensionValue] = ListMap.empty
)
```

Do you maybe have any ideas what could be the cause?

_Originally posted by @adamw in https://github.com/softwaremill/quicklens/issues/236#issuecomment-2355222436_

--- a/quicklens/src/main/scala-3/com/softwaremill/quicklens/QuicklensMacros.scala
+++ b/quicklens/src/main/scala-3/com/softwaremill/quicklens/QuicklensMacros.scala
@@ -168,13 +168,13 @@ object QuicklensMacros {
     def symbolAccessorByNameOrError(sym: Symbol, name: String): Symbol = {
       val mem = sym.fieldMember(name)
       if mem != Symbol.noSymbol then mem
-      else report.errorAndAbort(noSuchMember(sym.name, name))
+      else symbolMethodByNameOrError(sym, name)
     }
 
     def symbolMethodByNameOrError(sym: Symbol, name: String): Symbol = {
       sym.methodMember(name) match
         case List(m) => m
-        case Nil     => symbolAccessorByNameOrError(sym, name)
+        case Nil     => report.errorAndAbort(noSuchMember(sym.name, name))
         case _       => report.errorAndAbort(multipleMatchingMethods(sym.name, name))
     }
 
@@ -202,9 +202,10 @@ object QuicklensMacros {
         obj: Term,
         fields: Seq[(PathSymbol.Field, Seq[PathTree])]
     ): Term = {
-      val objSymbol = obj.tpe.widenAll.matchingTypeSymbol
+      val objTpe = obj.tpe.widenAll
+      val objSymbol = objTpe.matchingTypeSymbol
       if isSum(objSymbol) then {
-        obj.tpe.widenAll match {
+        objTpe match {
           case AndType(_, _) =>
             report.errorAndAbort(
               s"Implementation limitation: Cannot modify sealed hierarchies mixed with & types. Try providing a more specific type."
@@ -235,7 +236,7 @@ object QuicklensMacros {
       } else if isProduct(objSymbol) || isProductLike(objSymbol) then {
         val copy = symbolMethodByNameOrError(objSymbol, "copy")
         val argsMap: Map[String, Term] = fields.map { (field, trees) =>
-          val fieldMethod = symbolMethodByNameOrError(objSymbol, field.name)
+          val fieldMethod = symbolAccessorByNameOrError(objSymbol, field.name)
           val resTerm: Term = trees.foldLeft[Term](Select(obj, fieldMethod)) { (term, tree) =>
             mapToCopy(owner, mod, term, tree)
           }
@@ -243,7 +244,7 @@ object QuicklensMacros {
           field.name -> namedArg
         }.toMap
 
-        val typeParams = obj.tpe.widenAll match {
+        val typeParams = objTpe match {
           case AppliedType(_, typeParams) => Some(typeParams)
           case _                          => None
         }
@@ -272,7 +273,7 @@ object QuicklensMacros {
           case _                => Apply(Select(obj, copy), args)
         }
       } else
-        report.errorAndAbort(s"Unsupported source object: must be a case class or sealed trait, but got: $objSymbol")
+        report.errorAndAbort(s"Unsupported source object: must be a case class or sealed trait, but got: $objSymbol of type ${objTpe.show} (${obj.show})")
     }
 
     def applyFunctionDelegate(

--- a/quicklens/src/test/scala-3/com/softwaremill/quicklens/test/ExplicitCopyTest.scala
+++ b/quicklens/src/test/scala-3/com/softwaremill/quicklens/test/ExplicitCopyTest.scala
@@ -22,4 +22,18 @@ class ExplicitCopyTest extends AnyFlatSpec with Matchers {
     modified.show shouldEqual expected.show
   }
 
+  it should "modify a class that has a method with the same name as a field" in {
+    final case class PathItem()
+    final case class Paths(
+      pathItems: Map[String, PathItem] = Map.empty
+    )
+    final case class Docs(
+      paths: Paths = Paths()
+    ) {
+      def paths(paths: Paths): Docs = copy(paths = paths)
+    }
+    val docs = Docs()
+    docs.modify(_.paths.pathItems).using(m => m + ("a" -> PathItem()))
+  }
+
 }

